File: go/protein-translation/protein_translation.go
package protein

const testVersion = 1

type polypeptideTranslation struct {
	codons      []string
	polypeptide string
}

var polypeptideTranslations = []polypeptideTranslation{
	{[]string{"AUG"}, "Methionine"},
	{[]string{"UUU", "UUC"}, "Phenylalanine"},
	{[]string{"UUA", "UUG"}, "Leucine"},
	{[]string{"UCU", "UCC", "UCA", "UCG"}, "Serine"},
	{[]string{"UAU", "UAC"}, "Tyrosine"},
	{[]string{"UGU", "UGC"}, "Cysteine"},
	{[]string{"UGG"}, "Tryptophan"},
	{[]string{"UAA", "UAG", "UGA"}, "STOP"},
}

// FromCodon converts a codon to a polypeptide.
func FromCodon(sequence string) string {
	for _, t := range polypeptideTranslations {
		for _, codon := range t.codons {
			if sequence == codon {
				return t.polypeptide
			}
		}
	}

	return ""
}

// FromRNA translates an RNA sequence into a protein.
func FromRNA(sequence string) (protein []string) {
	for i := 0; i < len(sequence); i += 3 {
		polypeptide := FromCodon(sequence[i : i+3])
		if polypeptide == "STOP" {
			break
		}
		protein = append(protein, polypeptide)
	}

	return
}

File: haskell/protein-translation/src/ProteinTranslation.hs
module ProteinTranslation(proteins) where

proteins :: String -> Maybe [String]
proteins s = Just (codons s)

codons :: String -> [String]
codons [] = []
codons s = build (protein (take 3 s)) (drop 3 s)

build :: String -> String -> [String]
build "STOP" _ = []
build s rest = s : codons rest

protein :: String -> String
protein s = case s of
    "AUG" -> "Methionine"
    "UUU" -> "Phenylalanine"
    "UUC" -> "Phenylalanine"
    "UUA" -> "Leucine"
    "UUG" -> "Leucine"
    "UCU" -> "Serine"
    "UCC" -> "Serine"
    "UCA" -> "Serine"
    "UCG" -> "Serine"
    "UAU" -> "Tyrosine"
    "UAC" -> "Tyrosine"
    "UGU" -> "Cysteine"
    "UGC" -> "Cysteine"
    "UGG" -> "Tryptophan"
    "UAA" -> "STOP"
    "UAG" -> "STOP"
    "UGA" -> "STOP"
    _     -> error "Invalid codon"
